### DISCUSSION TITLE
Query results as array of arrays?

### DISCUSSION THREAD

**Discussion by ozum** [prisma/prisma#17554] [Q&A]
Hi,

## TLDR;

Having data as an array of arrays is much more efficient for some use cases. Is it possible using the Prisma ORM?

## What?

Prisma, like most ORM's returns query results as array of objects:

`prisma.product.findMany();`

```js
[
  { id: 1, color: "red", ... , isActive: true },
  { id: 2, color: "green", ..., isActive: true },
]
```

I want to have results as array of arrays:

```js
[
  [1, "red", ...., true],
  [2, "green", ...., true],
]
```

For example, node-postgres has [row-mode](https://node-postgres.com/features/queries#row-mode), Perl DBI has [selectall_arrayref()](https://metacpan.org/pod/DBD::Pg#selectall_arrayref) method for this.

## Rationale:

For large data sets an array of objects "[{}, {}]" has some disadvantages, because of unnecessary column names repeating in each object.

* Higher data size transferred between server and client.
* Higher memory consumption in the client.
* Higher CPU time to process large data set. (Maybe!!! This depends of further processing the data nad usage model)

## Existing Optimizations:

I'm very well aware of existing optimizations:

✅ Compression (gzip, brotli, etc.) works very well for repeating texts which reduce data transferred significantly.
  * Takes extra CPU time.
  * Still worse than an array of arrays. The best compression is the lack of data.

✅ V8 memory optimizations.
  * V8 cannot be sure all current data (db rows) and future data (db rows) are in the same structure.

## Workarounds

* Convert data to an array of arrays on the server manually or using a middleware.

DBMS already presents data as an array of arrays (or equivalent). DB drivers usually process the data to convert it to an array of objects. Converting it again to an array of arrays is redundant work. Even maybe, sometimes it will be converted to an array of objects again.

## Further Exploration

How to handle & use data on the client side efficiently and as much as developer-friendly supporting TypeScript? This can be a thought exercise...

* Classes

```ts
class Product {
  private rowData: { id: number, name: string, .... };
  
  get id() {
    return this.rowData[0]
  }

  get name() {
    return this.rowData[2]
  }
}
```

* A class extending Array

```ts
class KeyedArray<T> extends Array {
  constructor(keyList: string[], data: T[] ) {
  }
}

class Product<T> extends KeyedArray<T> {

}
```

* Accessing arrays using const variables:

```js
const id = 0;
const color = 1;
const name = 2;

data.forEach((row) => {
  row[color]
});
```

**Comment by nurul3101** [comment]
Hello @ozum 👋 

For now, Prisma doesn't support returning results as an array of arrays.
You could use [Prisma Client Extension to modify the result set and convert the objects to arrays before the response is sent to the client.](https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions/query#add-or-modify-a-custom-query-for-all-operations-in-a-specific-model)

Can you create a [Feature Request](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) for adding native support for sending responses in arrays instead of objects?

**Reply by ozum** [reply]
Created #17576

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
